### PR TITLE
Feat/issue 5376 aggregate subtask costs

### PR DESCRIPTION
**Pull request by taltas**
### Related GitHub Issue

- https://github.com/RooCodeInc/Roo-Code/issues/5376

Closes: #5376

### Description

When using Orchestrator (boomerang) mode, parent tasks delegate work to subtasks, but the parent task’s cost display previously reflected only its own API usage. This change adds a recursive aggregation layer so parent tasks can display **cumulative** cost across their entire subtask tree.

Key changes:
- Core aggregation logic: adds a recursive cost aggregator with cycle protection and missing-task handling.
- Extension ↔ webview bridge: exposes an aggregated-cost request so the UI can fetch totals for a given task.
- UI: requests aggregated costs for tasks with `childIds` and displays the aggregated total (with breakdown).
- Tests: adds coverage for aggregation logic.

Screenshot 2026-01-15 at 1 41 12 pm

https://github.com/user-attachments/assets/793eb805-e17d-4b98-8638-f4c318dd1540

### Test Procedure

1. Unit tests:
   - `cd packages/core && npx vitest run src/message-utils/__tests__/aggregateTaskCosts.spec.ts`
2. Typecheck:
   - `cd packages/core && pnpm run check-types`
3. Lint:
   - `cd packages/core && pnpm run lint`
   - `cd webview-ui && pnpm run lint`
4. Manual sanity check (UI):
   - Create an Orchestrator task with subtasks and confirm the parent task header cost reflects the sum of parent + subtasks.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

N/A

### Documentation Updates

None.

### Additional Notes

- Includes cycle detection to avoid infinite recursion.
- Missing child tasks are treated as zero-cost (with a warning).

### Get in Touch

@taltas

----

> [!IMPORTANT]
> Adds recursive cost aggregation for tasks and subtasks, updating backend logic and UI to display cumulative costs with detailed breakdowns.
> 
>   - **Behavior**:
>     - Adds recursive cost aggregation logic in `aggregateTaskCosts.ts` with cycle protection and missing-task handling.
>     - Updates `webviewMessageHandler.ts` to handle `getTaskWithAggregatedCosts` messages, fetching and posting aggregated costs.
>     - Modifies `ChatView.tsx` to request and display aggregated costs when tasks change.
>   - **UI**:
>     - Updates `TaskHeader.tsx` to display aggregated costs and breakdowns if subtasks exist.
>     - Adds tooltips in `TaskHeader.tsx` to show detailed cost breakdowns.
>   - **Tests**:
>     - Adds `aggregateTaskCosts.spec.ts` to test cost aggregation logic, including edge cases like circular references and missing tasks.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for f49ef4132b4cc78d197eac88c0eb2cb5e77d3516. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by ghost** [comment]
> **Rooviewer**  Clock &nbsp;&nbsp;[See task on Roo Cloud](https://app.roocode.com/cloud-jobs/5cJ1Mw3m?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review.sync)

All issues addressed. Ready to merge.

- [x] Remove debug logging statements (`[Aggregated Costs]` console.log calls) from production code

> <details>
> <summary>Previous reviews</summary>
> 
> - fe0c1ab: [Review #1](https://app.roocode.com/cloud-jobs/QJCUdAnf?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 17975eb: [Review #2](https://app.roocode.com/cloud-jobs/iP8oXox1?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - d7f4d7c: [Review #3](https://app.roocode.com/cloud-jobs/CKUTN06m?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - f49ef41: [Review #4](https://app.roocode.com/cloud-jobs/sFS0jA0d?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - f49ef41: [Review #5](https://app.roocode.com/cloud-jobs/BN9wlA3W?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 00ad92d: [Review #6](https://app.roocode.com/cloud-jobs/PCOVBCct?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> </details>
<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=10757&commentId=3756333882).</sub>

File: packages/core/src/message-utils/aggregateTaskCosts.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+export async function aggregateTaskCostsRecursive(
+	taskId: string,
+	getTaskHistory: (id: string) => Promise<HistoryItem | undefined>,
+	visited: Set<string> = new Set(),
+): Promise<AggregatedCosts> {
+	console.log("[Aggregated Costs][CORE] aggregateTaskCostsRecursive enter", {
+		taskId,
+		visitedSize: visited.size,
+		visited: Array.from(visited),
+	})
```
Comment (ghost) [inline packages/core/src/message-utils/aggregateTaskCosts.ts]: Debug logging statements with the `[Aggregated Costs]` prefix should be removed before merging to production. These appear throughout the codebase:

- This file (`aggregateTaskCosts.ts`): lines 26-30, 33-34, 42-43, 50-54, 59-62, 68-72, 85-90
- `webviewMessageHandler.ts`: lines 790-793, 798-802, 811-813
- `ChatView.tsx`: lines 218-223, 501-508, 928-943
- `TaskHeader.tsx`: lines 74-82

Debug logs in production pollute the console, can expose task/cost information, and add unnecessary performance overhead.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10757&fixId=64e0c82f3f) or mention @roomote and request a fix.</sub>

**Comment by taltas** [comment]
Addressed review comment re: debug logging.

- Removed all `[Aggregated Costs]` debug `console.log` statements from:
  - [`aggregateTaskCostsRecursive()`](packages/core/src/message-utils/aggregateTaskCosts.ts:21)
  - [`webviewMessageHandler`](src/core/webview/webviewMessageHandler.ts:787)
  - [`ChatView`](webview-ui/src/components/chat/ChatView.tsx:1)
  - [`TaskHeader`](webview-ui/src/components/chat/TaskHeader.tsx:54)

Change is in commit `f49ef4132` on branch `feat/issue-5376-aggregate-subtask-costs`.

**Comment by ghost** [comment]
> **Rooviewer**  Clock &nbsp;&nbsp;[See task on Roo Cloud](https://app.roocode.com/cloud-jobs/sFS0jA0d?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review)

All issues addressed. Ready to merge.

- [x] Remove debug logging statements (`[Aggregated Costs]` console.log calls) from production code

<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=10757&commentId=3756399944).</sub>

File: packages/core/package.json [hunk 2]
```diff
@@ -6,7 +6,8 @@
 	"exports": {
 		".": "./src/index.ts",
 		"./cli": "./src/cli.ts",
-		"./browser": "./src/browser.ts"
+		"./browser": "./src/browser.ts",
+		"./message-utils/aggregateTaskCosts": "./src/message-utils/aggregateTaskCosts.ts"
```
Comment (mrubens) [inline packages/core/package.json]: Mind explaining what this change does?

Comment (taltas) [reply packages/core/package.json]: Yeh, this was the agent being dumb 😅 I just fixed this.

File: packages/core/src/message-utils/aggregateTaskCosts.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * Get aggregated costs for display, handling incomplete tasks gracefully.
+ */
+export async function getDisplayCosts(
+	taskId: string,
+	getTaskHistory: (id: string) => Promise<HistoryItem | undefined>,
+): Promise<{
+	displayCost: number
+	showAggregated: boolean
```
Comment (mrubens) [inline packages/core/src/message-utils/aggregateTaskCosts.ts]: Super nitpick, but maybe instead of needing `showAggregated` you could just check if there's a `breakdown`?

File: packages/core/src/message-utils/aggregateTaskCosts.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	breakdown?: string
+}> {
+	const aggregated = await aggregateTaskCostsRecursive(taskId, getTaskHistory)
+
+	const hasChildren = aggregated.childrenCost > 0
+	const displayCost = aggregated.totalCost
+
+	let breakdown: string | undefined
+	if (hasChildren) {
+		breakdown = `Own: $${aggregated.ownCost.toFixed(2)} + Subtasks: $${aggregated.childrenCost.toFixed(2)}`
```
Comment (mrubens) [inline packages/core/src/message-utils/aggregateTaskCosts.ts]: In the extension we internationalize all user-facing strings (there's a `translate` mode that's convenient for it in the Roo-Code repo)

File: webview-ui/src/components/chat/ChatView.tsx [hunk 5]
```diff
@@ -207,6 +217,8 @@ const ChatViewComponent: React.ForwardRefRenderFunction<ChatViewRef, ChatViewPro
 		}
 	}, [])
 
+	// (intentionally no debug logging)
```
Comment (mrubens) [inline webview-ui/src/components/chat/ChatView.tsx]: Probably don't need this?

File: webview-ui/src/components/chat/ChatView.tsx [hunk 6]
```diff
@@ -467,6 +479,18 @@ const ChatViewComponent: React.ForwardRefRenderFunction<ChatViewRef, ChatViewPro
 		userRespondedRef.current = false
 	}, [task?.ts])
 
+	const taskTs = task?.ts
```
Comment (mrubens) [inline webview-ui/src/components/chat/ChatView.tsx:481]: Is this updated on every task message?

File: webview-ui/src/components/chat/ChatView.tsx [hunk 7]
```diff
@@ -1438,6 +1471,28 @@ const ChatViewComponent: React.ForwardRefRenderFunction<ChatViewRef, ChatViewPro
...
+								aggregatedCostsMap.has(currentTaskItem.id) &&
+								aggregatedCostsMap.get(currentTaskItem.id)!.childrenCost > 0
+							)
+						}
+						costBreakdown={
+							currentTaskItem?.id && aggregatedCostsMap.has(currentTaskItem.id)
+								? (() => {
+										const costs = aggregatedCostsMap.get(currentTaskItem.id)!
+										return costs.childrenCost > 0
+											? `Own: $${costs.ownCost.toFixed(2)} + Subtasks: $${costs.childrenCost.toFixed(2)}`
```
Comment (mrubens) [inline webview-ui/src/components/chat/ChatView.tsx]: Could we centralize the logic for generating this string (maybe reusing the logic from earlier in the PR?)

File: webview-ui/src/components/chat/ChatView.tsx (hunk 6 above)
Comment (taltas) [reply webview-ui/src/components/chat/ChatView.tsx:481]: No, it should be updated on switching to a different task or resuming a task only.

**Review by mrubens — APPROVED** [review]
